### PR TITLE
NMS-10264: Only increment counter if alarm type is not of type resolution

### PR DESCRIPTION
**Pull request by pbrane**
Thanks for taking time to contribute!

Please read our [Contribution Guidelines](https://github.com/OpenNMS/opennms/blob/develop/CONTRIBUTING.md) and format the title of the pull request in the format of:

${JIRA-ISSUE-NUMBER}: a-short-subject-title

Please use the [JIRA](https://issues.opennms.org) issue number and create a link in the JIRA issue back to this pull request so we have a quick reference from the issue to the pull request.

* JIRA: http://issues.opennms.org/browse/${JIRA-ISSUE-NUMBER}

Our [continuous integration system](http://bamboo.internal.opennms.com:8085) will test and verify your changes.

### CODE REVIEW

File: opennms-alarms/daemon/src/main/java/org/opennms/netmgt/alarmd/AlarmPersisterImpl.java [hunk 1]
```diff
@@ -154,7 +154,10 @@ private void reduceEvent(OnmsEvent e, OnmsAlarm alarm, Event event) {
         // Always set these
         alarm.setLastEvent(e);
         alarm.setLastEventTime(e.getEventTime());
-        alarm.setCounter(alarm.getCounter() + 1);
+        
+        if (event.getAlarmData().getAlarmType().equals(alarm.getAlarmType())) {
```
Comment (j-white) [inline opennms-alarms/daemon/src/main/java/org/opennms/netmgt/alarmd/AlarmPersisterImpl.java]: I'd change this to ` if (Objects.equals(event.getAlarmData().getAlarmType(), alarm.getAlarmType())) {` to help guard against NPEs.

Comment (pbrane) [reply opennms-alarms/daemon/src/main/java/org/opennms/netmgt/alarmd/AlarmPersisterImpl.java]: Sure.

**Comment by pbrane** [comment]
When we change the configuration to have a single alarm vs. the current default way of having a "clearing alarm", the counter increments when the clearing event is reduced to the alarm.  We should only increment the counter if the alarm repeats.  So, if the alarm-type changes, we shouldn't increment that counter.

I also need to verify that the alarm-type gets updated on reduction.  I'll do that with the NPE guards you suggested.

**Comment by pbrane** [comment]
https://issues.opennms.org/browse/NMS-10264

**Comment by j-white** [comment]
Just to validate the current logic:
* If an alarm has type 1, and we send another related event with type 2, we want to change the alarm's type from 1 to 2?
* If an alarm has type 2, and we send multiple events, would we not want to increment the count for each of these?

**Comment by pbrane** [comment]
If we send a clearing event (alarm type 2), we update the alarm type, logmsg, and not the count.  If the problem reoccurs (alarm type 1), we increment the counter, update the logmsg and set the alarm type to back to problem (alarm type 1).

**Comment by j-white** [comment]
With these changes, if we send 2 clear events the count will always be 1.

Can you clarify why we want to change the alarm type on reduction? Shouldn't we use the severity to track whether or not an alarm has "cleared"?

**Comment by RangerRick** [comment]
You're gonna have to rename this branch to something shorter if you want it to pass in bamboo. 

`... /file-----home-bamboo-ci-bamboo-xml-data-build-dir-OPENNMS-ONMS2316-RPM-opennms-source-target-rpm-BUILD-opennms-23.0.0-0.20180716.onms2316.features.drools4alarms.betteralarmcounthandling.2-features-ticketing-otrs-integration-src-main-wsdl-TicketService.wsdl (File name too long)`

**Comment by pbrane** [comment]
We should only be tracking counts of problems.  When we move to having a single alarm vs. the current crappy two alarm situation, this makes more sense.

File: opennms-alarms/daemon/src/main/java/org/opennms/netmgt/alarmd/AlarmPersisterImpl.java [hunk 2]
```diff
@@ -154,74 +155,87 @@ private void reduceEvent(OnmsEvent e, OnmsAlarm alarm, Event event) {
         // Always set these
         alarm.setLastEvent(e);
         alarm.setLastEventTime(e.getEventTime());
-        alarm.setCounter(alarm.getCounter() + 1);
 
-        if (!event.getAlarmData().hasUpdateFields()) {
+        AlarmData alarmData = event.getAlarmData();
+        if (alarmData != null) {
```
Comment (j-white) [inline opennms-alarms/daemon/src/main/java/org/opennms/netmgt/alarmd/AlarmPersisterImpl.java]: The event will be dropped earlier if this is null, so there's no need to check this again here - see  `checkEventSanityAndDoWeProcess`

File: opennms-alarms/daemon/src/main/java/org/opennms/netmgt/alarmd/AlarmPersisterImpl.java [hunk 3]
```diff
@@ -154,74 +155,87 @@ private void reduceEvent(OnmsEvent e, OnmsAlarm alarm, Event event) {
...
-                if (fieldName.equalsIgnoreCase("LogMsg") && !field.isUpdateOnReduction()) {
-                    continue;
+            Integer newAlarmType = alarmData.getAlarmType();
+            Integer currentAlarmType = alarm.getAlarmType();
+            if (newAlarmType != null && currentAlarmType != null) {
+                if (!newAlarmType.equals(OnmsAlarm.RESOLUTION_TYPE)) {
+                    alarm.setCounter(alarm.getCounter() + 1);
                 } else {
-                    alarm.setLogMsg(e.getEventLogMsg());
+                    alarm.setAlarmType(newAlarmType);
```
Comment (j-white) [inline opennms-alarms/daemon/src/main/java/org/opennms/netmgt/alarmd/AlarmPersisterImpl.java]: It looks like an alarm can be updated from type 1 to type 2, but it won't get updated back to type 1 if the problem reoccurs.

File: opennms-alarms/daemon/src/main/java/org/opennms/netmgt/alarmd/AlarmPersisterImpl.java (hunk 2 above)
Comment (pbrane) [reply opennms-alarms/daemon/src/main/java/org/opennms/netmgt/alarmd/AlarmPersisterImpl.java]: But asked in previous code change to do Null checking on these same objects.

File: opennms-alarms/daemon/src/main/java/org/opennms/netmgt/alarmd/AlarmPersisterImpl.java (hunk 3 above)
Comment (pbrane) [reply opennms-alarms/daemon/src/main/java/org/opennms/netmgt/alarmd/AlarmPersisterImpl.java]: Fixed in latest commit.

File: opennms-alarms/daemon/src/main/java/org/opennms/netmgt/alarmd/AlarmPersisterImpl.java (hunk 2 above)
Comment (j-white) [reply opennms-alarms/daemon/src/main/java/org/opennms/netmgt/alarmd/AlarmPersisterImpl.java]: I had recommended changing the line in question to `if (Objects.equals(event.getAlarmData().getAlarmType(), alarm.getAlarmType())) {`.

Using `Objects.equals` in this case would only help prevent an NPE if/when the first call to `getAlarmType()` returned a null value.

**Comment by RangerRick** [comment]
FYI, over on [the other version of this branch](https://bamboo.opennms.org/browse/OPENNMS-ONMS2321-UNIT-2/log) (that passes the wsdl length thing) there's currently a test failure:

```
17-Jul-2018 10:31:26 | Failed tests:
17-Jul-2018 10:31:26 | org.opennms.netmgt.alarmd.AlarmdBlackboxIT.canFlapAlarm(org.opennms.netmgt.alarmd.AlarmdBlackboxIT)
17-Jul-2018 10:31:26 | Run 1: AlarmdBlackboxIT.canFlapAlarm:161
17-Jul-2018 10:31:26 | Expected: <2>
17-Jul-2018 10:31:26 | but: was <1>
17-Jul-2018 10:31:26 | Run 2: AlarmdBlackboxIT.canFlapAlarm:161
17-Jul-2018 10:31:26 | Expected: <2>
17-Jul-2018 10:31:26 | but: was <1>
```

**Pull request closed without merging**